File: app/src/main/java/com/angogasapps/myfamily/ui/screens/buy_list/adapters/ProductsAdapter.java
package com.angogasapps.myfamily.ui.screens.buy_list.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.angogasapps.myfamily.R;
import com.angogasapps.myfamily.databinding.ProductInBuyListHolderBinding;
import com.angogasapps.myfamily.models.buy_list.BuyList;
import com.angogasapps.myfamily.models.buy_list.BuyListEvent;
import com.angogasapps.myfamily.ui.screens.buy_list.dialogs.ChangeOrDeleteProductDialog;
import com.angogasapps.myfamily.utils.BuyListUtils;

public class ProductsAdapter extends RecyclerView.Adapter<ProductsAdapter.ProductHolder>{
    private Context context;
    private BuyList buyList;
    private LayoutInflater inflater;

    public ProductsAdapter(Context context, BuyList buyList) {
        inflater = LayoutInflater.from(context);
        this.context = context;
        this.buyList = buyList;
    }

    @NonNull
    @Override
    public ProductHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ProductHolder(
                ProductInBuyListHolderBinding.bind(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.product_in_buy_list_holder, parent, false))
        );

//        return new ProductHolder(ProductInBuyListHolderBinding.inflate(inflater));
    }

    @Override
    public void onBindViewHolder(@NonNull ProductHolder holder, int position) {
        BuyList.Product product = buyList.getProducts().get(position);

        holder.binding.textName.setText(product.getName());
        holder.binding.commentText.setText(BuyListUtils.getCorrectComment(product));

        holder.binding.getRoot().setOnLongClickListener(v -> {
            new ChangeOrDeleteProductDialog(context, buyList.getId(), product).show();
            return true;
        });

        holder.binding.commentText.setVisibility(product.getComment().equals("")? View.GONE : View.VISIBLE);
    }

    @Override
    public int getItemCount() {
        return buyList.getProducts().size();
    }


    public void update(BuyListEvent event){
        if (event.getEvent().equals(BuyListEvent.EBuyListEvents.productAdded)){
            this.notifyItemChanged(event.getIndex());
        }else if(event.getEvent().equals(BuyListEvent.EBuyListEvents.productRemoved)){
            this.notifyItemRemoved(event.getIndex());
        }else if(event.getEvent().equals(BuyListEvent.EBuyListEvents.productChanged)){
            this.notifyItemChanged(event.getIndex());
        }
    }

    public static class ProductHolder extends RecyclerView.ViewHolder{
        ProductInBuyListHolderBinding binding;

        public ProductHolder(@NonNull ProductInBuyListHolderBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

}

File: app/src/main/java/com/angogasapps/myfamily/ui/screens/buy_list/adapters/ProductsAdapter.kt
package com.angogasapps.myfamily.ui.screens.buy_list.adapters

import android.content.Context
import com.angogasapps.myfamily.models.buy_list.BuyList
import androidx.recyclerview.widget.RecyclerView
import com.angogasapps.myfamily.ui.screens.buy_list.adapters.ProductsAdapter.ProductHolder
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.angogasapps.myfamily.R
import com.angogasapps.myfamily.models.buy_list.BuyList.Product
import com.angogasapps.myfamily.utils.BuyListUtils
import android.view.View.OnLongClickListener
import com.angogasapps.myfamily.databinding.ProductInBuyListHolderBinding
import com.angogasapps.myfamily.ui.screens.buy_list.dialogs.ChangeOrDeleteProductDialog
import com.angogasapps.myfamily.models.buy_list.BuyListEvent

class ProductsAdapter(private val context: Context, private val buyList: BuyList) : RecyclerView.Adapter<ProductHolder>() {
    private val inflater: LayoutInflater = LayoutInflater.from(context)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductHolder {
        return ProductHolder(
            ProductInBuyListHolderBinding.bind(
                LayoutInflater.from(parent.context)
                    .inflate(R.layout.product_in_buy_list_holder, parent, false)
            )
        )
    }

    override fun onBindViewHolder(holder: ProductHolder, position: Int) {
        val product = buyList.products[position]
        holder.binding.textName.text = product.name
        holder.binding.commentText.text = BuyListUtils.getCorrectComment(product)
        holder.binding.root.setOnLongClickListener {
            ChangeOrDeleteProductDialog(context, buyList.id, product).show()
            true
        }
        holder.binding.commentText.visibility =
            if (product.comment == "") View.GONE else View.VISIBLE
    }

    override fun getItemCount(): Int {
        return buyList.products.size
    }

    fun update(event: BuyListEvent) {
        when (event.event) {
            BuyListEvent.EBuyListEvents.productAdded -> {
                this.notifyItemChanged(event.index)
            }
            BuyListEvent.EBuyListEvents.productRemoved -> {
                notifyItemRemoved(event.index)
            }
            BuyListEvent.EBuyListEvents.productChanged -> {
                this.notifyItemChanged(event.index)
            }
        }
    }

    class ProductHolder(var binding: ProductInBuyListHolderBinding)
        : RecyclerView.ViewHolder(binding.root)

}